File: remote/src/main/java/com/bizard/homesmokeremote/SystemStatusActivity.java
package com.bizard.homesmokeremote;

import android.Manifest;
import android.app.Activity;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.Gravity;
import android.view.View;
import android.view.WindowInsets;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import org.json.JSONArray;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/** Read-only diagnostics assembled from Remote runtime state and local stores. */
public final class SystemStatusActivity extends Activity {
    private static final int NAVY=Color.rgb(9,47,73),BG=Color.rgb(245,247,250),CARD=Color.WHITE,TEXT=Color.rgb(21,31,47),MUTED=Color.rgb(101,116,139),BORDER=Color.rgb(220,225,232),GREEN=Color.rgb(35,151,83),ORANGE=Color.rgb(231,138,7);
    private static final long LIVE_MS=10000L,HISTORY_MS=24L*60L*60L*1000L;
    private SharedPreferences prefs;
    private TelemetryHistoryStore telemetry;
    private OperationalHistoryStore ops;
    private TextView connectionBody,telemetryBody,historyBody,notificationsBody,remoteBody;
    private final Handler handler=new Handler(Looper.getMainLooper());
    private final Runnable tick=new Runnable(){@Override public void run(){refresh();handler.postDelayed(this,1000L);}};

    @Override protected void onCreate(Bundle state){
        super.onCreate(state);
        prefs=getSharedPreferences("homesmoke_remote",MODE_PRIVATE);
        telemetry=new TelemetryHistoryStore(this);
        ops=new OperationalHistoryStore(this);
        View root=buildRoot();setContentView(root);applyInsets(root);refresh();handler.postDelayed(tick,1000L);
    }

    @Override protected void onDestroy(){handler.removeCallbacks(tick);if(telemetry!=null)telemetry.close();if(ops!=null)ops.close();super.onDestroy();}

    private View buildRoot(){
        LinearLayout root=new LinearLayout(this);root.setOrientation(LinearLayout.VERTICAL);root.setBackgroundColor(BG);
        root.addView(buildBar(),new LinearLayout.LayoutParams(-1,dp(60)));
        ScrollView scroll=new ScrollView(this);scroll.setFillViewport(true);scroll.setVerticalScrollBarEnabled(false);
        LinearLayout page=new LinearLayout(this);page.setOrientation(LinearLayout.VERTICAL);page.setPadding(dp(4),dp(4),dp(4),dp(18));page.setBackgroundColor(BG);
        LinearLayout intro=card();intro.addView(text("Состояние системы",18,true,TEXT));intro.addView(detail("Только диагностика Remote. Здесь нет новых команд или полей протокола."));page.addView(intro,margin(8,4,8,4));
        connectionBody=addStatusCard(page,"Связь");telemetryBody=addStatusCard(page,"Телеметрия");historyBody=addStatusCard(page,"Локальная история");notificationsBody=addStatusCard(page,"Уведомления");remoteBody=addStatusCard(page,"Remote");
        scroll.addView(page,new ScrollView.LayoutParams(-1,-2));root.addView(scroll,new LinearLayout.LayoutParams(-1,0,1));return root;
    }

    private View buildBar(){
        LinearLayout bar=new LinearLayout(this);bar.setGravity(Gravity.CENTER_VERTICAL);bar.setPadding(dp(8),0,dp(10),0);bar.setBackgroundColor(NAVY);
        TextView back=text("‹",34,false,Color.WHITE);back.setGravity(Gravity.CENTER);back.setOnClickListener(v->finish());bar.addView(back,new LinearLayout.LayoutParams(dp(44),dp(46)));
        LinearLayout titles=new LinearLayout(this);titles.setOrientation(LinearLayout.VERTICAL);titles.setGravity(Gravity.CENTER_VERTICAL);titles.addView(text("Состояние системы",18,true,Color.WHITE));titles.addView(text("Диагностика HomeSmoke Remote",11,false,Color.rgb(211,222,232)));bar.addView(titles,new LinearLayout.LayoutParams(0,-1,1));return bar;
    }

    private TextView addStatusCard(LinearLayout page,String title){
        LinearLayout c=card();c.addView(text(title,18,true,TEXT));TextView body=detail("Обновление…");body.setTextColor(TEXT);body.setPadding(0,dp(7),0,0);c.addView(body);page.addView(c,margin(8,4,8,4));return body;
    }

    private void refresh(){
        if(prefs==null||telemetry==null||ops==null)return;
        long now=System.currentTimeMillis(),diagAt=prefs.getLong("diag_updated_at",0L),diagAge=diagAt>0?Math.max(0,now-diagAt):Long.MAX_VALUE;
        boolean diagFresh=diagAge<=5000L,test=prefs.getBoolean("test_mode_active",false)||prefs.getBoolean("diag_test_running",false);
        boolean mqttConnected=diagFresh&&prefs.getBoolean("diag_mqtt_connected",false),mqttConnecting=diagFresh&&prefs.getBoolean("diag_mqtt_connecting",false);
        String mqtt=mqttConnected?"Подключён":(mqttConnecting?"Подключение…":(diagFresh?"Отключён":"Нет актуального снимка"));
        String runtime=diagAt>0?"Снимок Remote · "+age(diagAt,now)+" назад":"Снимок Remote ещё не получен";
        connectionBody.setText("MQTT · "+mqtt+"\nИсточник данных · "+(test?"ТЕСТ · локальная симуляция":"Коптильня / MQTT")+"\n"+runtime);
        connectionBody.setTextColor(mqttConnected?GREEN:(test?ORANGE:(mqttConnecting?ORANGE:MUTED)));

        long latest=telemetry.latestTimestamp();boolean live=latest>0&&Math.abs(now-latest)<=LIVE_MS,latestTest=latest>0&&isTestTimestamp(latest);
        String device=prefs.getString("diag_device_id","—"),mode=prefs.getString("diag_mode_display","—"),auto=prefs.getString("diag_auto_display","—"),heater=prefs.getString("diag_heater_display","—");
        telemetryBody.setText("Последняя точка · "+(latest>0?dateTime(latest)+" · "+age(latest,now)+" назад":"нет данных")+"\nСвежесть · "+(live?"LIVE":"не свежая")+" · "+(latestTest?"тестовая запись":"обычные данные")+"\nУстройство · "+empty(device)+"\nРежим · "+empty(mode)+" · Auto "+empty(auto)+"\nТЭН · "+empty(heater));
        telemetryBody.setTextColor(live?(latestTest?ORANGE:GREEN):MUTED);

        int points=telemetry.countSamples(now-HISTORY_MS,now);List<OperationalHistoryStore.Session> sessions=ops.querySessions(1000);List<OperationalHistoryStore.Event> events=ops.queryEvents(500);OperationalHistoryStore.Session last=sessions.isEmpty()?null:sessions.get(0);
        String lastSession=last==null?"нет":last.title()+" · "+(last.active()?"активен":duration(last.durationMs()));
        historyBody.setText("Точки телеметрии за 24 ч · "+points+"\nСеансов сохранено · "+sessions.size()+"\nСобытий в журнале · "+events.size()+" / 500\nПоследний сеанс · "+lastSession+"\nСырые точки хранятся до 24 ч; сводки сеансов сохраняются отдельно.");historyBody.setTextColor(TEXT);

        boolean nConnection=prefs.getBoolean("notify_connection",true),nSet=prefs.getBoolean("notify_setpoint",true),nSession=prefs.getBoolean("notify_session",false);String permission;
        if(Build.VERSION.SDK_INT>=33)permission=checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)==PackageManager.PERMISSION_GRANTED?"разрешены системой":"нет системного разрешения";else permission="разрешение системы не требуется";
        notificationsBody.setText("Связь · "+onOff(nConnection)+"\nКамера достигла уставки · "+onOff(nSet)+"\nНачало/завершение сеанса · "+onOff(nSession)+"\nAndroid · "+permission);notificationsBody.setTextColor(permission.startsWith("нет")?ORANGE:TEXT);

        String command=prefs.getString("diag_command_display","—"),testName=prefs.getString("test_active_name","");
        remoteBody.setText("Версия · "+versionName()+"\nТестовый режим · "+(test?("активен"+(testName.isEmpty()?"":" · "+testName)):"выключен")+"\nПоследняя команда/статус · "+empty(command)+"\nAndroid · "+Build.VERSION.RELEASE+" (API "+Build.VERSION.SDK_INT+")");remoteBody.setTextColor(test?ORANGE:TEXT);
    }

    private boolean isTestTimestamp(long ts){
        if(prefs.getBoolean("test_mode_active",false)){long start=prefs.getLong("test_active_start",0L);if(start>0&&ts>=start)return true;}
        try{JSONArray a=new JSONArray(prefs.getString("test_intervals","[]"));for(int i=0;i<a.length();i++){JSONObject x=a.optJSONObject(i);if(x==null)continue;long s=x.optLong("start",0L),e=x.optLong("end",0L);if(s>0&&e>=s&&ts>=s&&ts<=e)return true;}}catch(Exception ignored){}
        return false;
    }

    private String versionName(){try{return getPackageManager().getPackageInfo(getPackageName(),0).versionName;}catch(Exception ignored){return "—";}}
    private static String onOff(boolean v){return v?"включено":"выключено";}
    private static String empty(String v){return v==null||v.trim().isEmpty()?"—":v.trim();}
    private static String dateTime(long ts){return new SimpleDateFormat("dd.MM HH:mm:ss",Locale.getDefault()).format(new Date(ts));}
    private static String age(long ts,long now){long sec=Math.max(0,(now-ts)/1000L);if(sec<60)return sec+" сек";long min=sec/60;if(min<60)return min+" мин";long h=min/60;if(h<24)return h+" ч "+(min%60)+" мин";return (h/24)+" д "+(h%24)+" ч";}
    private static String duration(long ms){long m=Math.max(0,ms/60000L),h=m/60,r=m%60;return h>0?h+" ч "+r+" мин":m+" мин";}

    private void applyInsets(View root){
        if(Build.VERSION.SDK_INT<21)return;
        root.setOnApplyWindowInsetsListener((view,i)->{int l,t,r,b;if(Build.VERSION.SDK_INT>=30){android.graphics.Insets x=i.getInsets(WindowInsets.Type.systemBars());l=x.left;t=x.top;r=x.right;b=x.bottom;}else{l=i.getSystemWindowInsetLeft();t=i.getSystemWindowInsetTop();r=i.getSystemWindowInsetRight();b=i.getSystemWindowInsetBottom();}view.setPadding(l,t,r,b);return i;});
        root.requestApplyInsets();getWindow().setStatusBarColor(NAVY);getWindow().setNavigationBarColor(BG);
    }

    private LinearLayout card(){LinearLayout c=new LinearLayout(this);c.setOrientation(LinearLayout.VERTICAL);c.setPadding(dp(12),dp(10),dp(12),dp(10));c.setBackground(roundStroke(CARD,18,BORDER,1));if(Build.VERSION.SDK_INT>=21)c.setElevation(dp(1));return c;}
    private TextView detail(String s){TextView t=text(s,12,false,MUTED);t.setLineSpacing(0,1.08f);return t;}
    private TextView text(String s,int sp,boolean bold,int color){TextView t=new TextView(this);t.setText(s);t.setTextSize(sp);t.setTextColor(color);if(bold)t.setTypeface(Typeface.DEFAULT_BOLD);return t;}
    private GradientDrawable round(int color,int radius){GradientDrawable g=new GradientDrawable();g.setColor(color);g.setCornerRadius(dp(radius));return g;}
    private GradientDrawable roundStroke(int color,int radius,int stroke,int width){GradientDrawable g=round(color,radius);g.setStroke(dp(width),stroke);return g;}
    private LinearLayout.LayoutParams margin(int l,int t,int r,int b){LinearLayout.LayoutParams p=new LinearLayout.LayoutParams(-1,-2);p.setMargins(dp(l),dp(t),dp(r),dp(b));return p;}
    private int dp(int v){return Math.round(v*getResources().getDisplayMetrics().density);}
}

File: remote/src/main/java/com/bizard/homesmokeremote/SystemStatusActivity.kt
package com.bizard.homesmokeremote

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.View
import android.view.WindowInsets
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import org.json.JSONArray
import org.json.JSONObject

/** Read-only diagnostics assembled from Remote runtime state and local stores. */
class SystemStatusActivity : Activity() {
    private var prefs: SharedPreferences? = null
    private var telemetry: TelemetryHistoryStore? = null
    private var ops: OperationalHistoryStore? = null
    private var connectionBody: TextView? = null
    private var telemetryBody: TextView? = null
    private var historyBody: TextView? = null
    private var notificationsBody: TextView? = null
    private var remoteBody: TextView? = null
    private val handler: Handler = Handler(Looper.getMainLooper())
    private val tick: Runnable =
        object : Runnable {
            public override fun run() {
                refresh()
                handler.postDelayed(this, 1000L)
            }
        }

    protected override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        prefs = getSharedPreferences("homesmoke_remote", Context.MODE_PRIVATE)
        telemetry = TelemetryHistoryStore(this)
        ops = OperationalHistoryStore(this)
        val root: View? = buildRoot()
        setContentView(root)
        applyInsets(root)
        refresh()
        handler.postDelayed(tick, 1000L)
    }

    protected override fun onDestroy() {
        handler.removeCallbacks(tick)
        if (telemetry != null) telemetry!!.close()
        if (ops != null) ops!!.close()
        super.onDestroy()
    }

    private fun buildRoot(): View? {
        val root: LinearLayout = LinearLayout(this)
        root.setOrientation(LinearLayout.VERTICAL)
        root.setBackgroundColor(BG)
        root.addView(buildBar(), LinearLayout.LayoutParams(-1, dp(60)))
        val scroll: ScrollView = ScrollView(this)
        scroll.setFillViewport(true)
        scroll.setVerticalScrollBarEnabled(false)
        val page: LinearLayout = LinearLayout(this)
        page.setOrientation(LinearLayout.VERTICAL)
        page.setPadding(dp(4), dp(4), dp(4), dp(18))
        page.setBackgroundColor(BG)
        val intro: LinearLayout? = card()
        intro!!.addView(text("Состояние системы", 18, true, TEXT))
        intro!!.addView(
            detail("Только диагностика Remote. Здесь нет новых команд или полей протокола.")
        )
        page.addView(intro, margin(8, 4, 8, 4))
        connectionBody = addStatusCard(page, "Связь")
        telemetryBody = addStatusCard(page, "Телеметрия")
        historyBody = addStatusCard(page, "Локальная история")
        notificationsBody = addStatusCard(page, "Уведомления")
        remoteBody = addStatusCard(page, "Remote")
        scroll.addView(page, android.widget.FrameLayout.LayoutParams(-1, -2))
        root.addView(scroll, LinearLayout.LayoutParams(-1, 0, 1f))
        return root
    }

    private fun buildBar(): View? {
        val bar: LinearLayout = LinearLayout(this)
        bar.setGravity(Gravity.CENTER_VERTICAL)
        bar.setPadding(dp(8), 0, dp(10), 0)
        bar.setBackgroundColor(NAVY)
        val back: TextView? = text("‹", 34, false, TEXT)
        back!!.setGravity(Gravity.CENTER)
        back!!.setOnClickListener({ v -> finish() })
        bar.addView(back, LinearLayout.LayoutParams(dp(44), dp(46)))
        val titles: LinearLayout = LinearLayout(this)
        titles.setOrientation(LinearLayout.VERTICAL)
        titles.setGravity(Gravity.CENTER_VERTICAL)
        titles.addView(text("Состояние системы", 18, true, TEXT))
        titles.addView(text("Диагностика HomeSmoke Remote", 11, false, MUTED))
        bar.addView(titles, LinearLayout.LayoutParams(0, -1, 1f))
        return bar
    }

    private fun addStatusCard(page: LinearLayout?, title: String?): TextView? {
        val c: LinearLayout? = card()
        c!!.addView(text(title, 18, true, TEXT))
        val body: TextView? = detail("Обновление…")
        body!!.setTextColor(TEXT)
        body!!.setPadding(0, dp(7), 0, 0)
        c!!.addView(body)
        page!!.addView(c, margin(8, 4, 8, 4))
        return body
    }

    private fun refresh() {
        if (prefs == null || telemetry == null || ops == null) return
        val now: Long = System.currentTimeMillis()
        val diagAt: Long = prefs!!.getLong("diag_updated_at", 0L)
        val diagAge: Long = if (diagAt > 0) Math.max(0, now - diagAt) else java.lang.Long.MAX_VALUE
        val diagFresh: Boolean = diagAge <= 5000L
        val test: Boolean =
            prefs!!.getBoolean("test_mode_active", false) ||
                prefs!!.getBoolean("diag_test_running", false)
        val mqttConnected: Boolean = diagFresh && prefs!!.getBoolean("diag_mqtt_connected", false)
        val mqttConnecting: Boolean = diagFresh && prefs!!.getBoolean("diag_mqtt_connecting", false)
        val mqtt: String =
            if (mqttConnected) "Подключён"
            else
                (if (mqttConnecting) "Подключение…"
                else (if (diagFresh) "Отключён" else "Нет актуального снимка"))
        val runtime: String? =
            if (diagAt > 0)
                "Диагностика обновлена · " +
                    (if (diagAge < 2000L) "сейчас" else age(diagAt, now)!! + " назад")
            else "Диагностика ещё не обновлялась"
        connectionBody!!.setText(
            "MQTT · " +
                mqtt +
                "\nИсточник данных · " +
                (if (test) "ТЕСТ · локальная симуляция" else "Коптильня / MQTT") +
                "\n" +
                runtime
        )
        connectionBody!!.setTextColor(
            if (mqttConnected) GREEN
            else (if (test) ORANGE else (if (mqttConnecting) ORANGE else MUTED))
        )

        val latest: Long = telemetry!!.latestTimestamp()
        val live: Boolean = latest > 0 && Math.abs(now - latest) <= LIVE_MS
        val latestTest: Boolean = latest > 0 && isTestTimestamp(latest)
        val device: String? = prefs!!.getString("diag_device_id", "—")
        val mode: String? = prefs!!.getString("diag_mode_display", "—")
        val auto: String? = prefs!!.getString("diag_auto_display", "—")
        val heater: String? = prefs!!.getString("diag_heater_display", "—")
        telemetryBody!!.setText(
            "Последняя точка · " +
                (if (latest > 0) dateTime(latest) + " · " + age(latest, now) + " назад"
                else "нет данных") +
                "\nСвежесть · " +
                (if (live) "LIVE" else "не свежая") +
                " · " +
                (if (latestTest) "тестовая запись" else "обычные данные") +
                "\nУстройство · " +
                empty(device) +
                "\nРежим · " +
                empty(mode) +
                " · Auto " +
                empty(auto) +
                "\nТЭН · " +
                empty(heater)
        )
        telemetryBody!!.setTextColor(if (live) (if (latestTest) ORANGE else GREEN) else MUTED)

        val points: Int = telemetry!!.countSamples(now - HISTORY_MS, now)
        val sessions: List<OperationalHistoryStore.Session?>? = ops!!.querySessions(1000)
        val events: List<OperationalHistoryStore.Event?>? = ops!!.queryEvents(500)
        val last: OperationalHistoryStore.Session? =
            if (sessions!!.isEmpty()) null else sessions!!.get(0)
        val lastSession: String? =
            if (last == null) "нет"
            else
                last!!.title() +
                    " · " +
                    (if (last!!.active()) "активен" else duration(last!!.durationMs()))
        historyBody!!.setText(
            "Точки телеметрии за 24 ч · " +
                points +
                "\nСеансов сохранено · " +
                sessions!!.size +
                "\nСобытий в журнале · " +
                events!!.size +
                " / 500\nПоследний сеанс · " +
                lastSession +
                "\nСырые точки хранятся до 24 ч; сводки сеансов сохраняются отдельно."
        )
        historyBody!!.setTextColor(TEXT)

        val nConnection: Boolean = prefs!!.getBoolean("notify_connection", true)
        val nSet: Boolean = prefs!!.getBoolean("notify_setpoint", true)
        val nSession: Boolean = prefs!!.getBoolean("notify_session", false)
        val permission: String?
        if (Build.VERSION.SDK_INT >= 33)
            permission =
                if (
                    checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) ==
                        PackageManager.PERMISSION_GRANTED
                )
                    "разрешены системой"
                else "нет системного разрешения"
        else permission = "разрешение системы не требуется"
        notificationsBody!!.setText(
            "Связь · " +
                onOff(nConnection) +
                "\nКамера достигла уставки · " +
                onOff(nSet) +
                "\nНачало/завершение сеанса · " +
                onOff(nSession) +
                "\nAndroid · " +
                permission
        )
        notificationsBody!!.setTextColor(if (permission!!.startsWith("нет")) ORANGE else TEXT)

        val command: String? = prefs!!.getString("diag_command_display", "—")
        val testName: String? = prefs!!.getString("test_active_name", "")
        remoteBody!!.setText(
            "Версия · " +
                versionName() +
                "\nТестовый режим · " +
                (if (test) ("активен" + (if (testName!!.isEmpty()) "" else " · " + testName!!))
                else "выключен") +
                "\nПоследняя команда/статус · " +
                empty(command) +
                "\nAndroid · " +
                Build.VERSION.RELEASE +
                " (API " +
                Build.VERSION.SDK_INT +
                ")"
        )
        remoteBody!!.setTextColor(if (test) ORANGE else TEXT)
    }

    private fun isTestTimestamp(ts: Long): Boolean {
        if (prefs!!.getBoolean("test_mode_active", false)) {
            val start: Long = prefs!!.getLong("test_active_start", 0L)
            if (start > 0 && ts >= start) return true
        }
        try {
            val a: JSONArray = JSONArray(prefs!!.getString("test_intervals", "[]"))
            for (i: Int in 0 until a.length()) {
                val x: JSONObject? = a.optJSONObject(i)
                if (x == null) continue
                val s: Long = x!!.optLong("start", 0L)
                val e: Long = x!!.optLong("end", 0L)
                if (s > 0 && e >= s && ts >= s && ts <= e) return true
            }
        } catch (ignored: Exception) {}

        return false
    }

    private fun versionName(): String? {
        try {
            return getPackageManager()!!.getPackageInfo(getPackageName(), 0)!!.versionName
        } catch (ignored: Exception) {
            return "—"
        }
    }

    private fun applyInsets(root: View?) {
        if (Build.VERSION.SDK_INT < 21) return
        root!!.setOnApplyWindowInsetsListener({ view, i ->
            val l: Int
            val t: Int
            val r: Int
            val b: Int
            if (Build.VERSION.SDK_INT >= 30) {
                val x: android.graphics.Insets? = i!!.getInsets(WindowInsets.Type.systemBars())
                l = x!!.left
                t = x!!.top
                r = x!!.right
                b = x!!.bottom
            } else {
                l = i!!.getSystemWindowInsetLeft()
                t = i!!.getSystemWindowInsetTop()
                r = i!!.getSystemWindowInsetRight()
                b = i!!.getSystemWindowInsetBottom()
            }
            view!!.setPadding(l, t, r, b)
            i
        })
        root!!.requestApplyInsets()
        getWindow()!!.setStatusBarColor(NAVY)
        getWindow()!!.setNavigationBarColor(BG)
        if (Build.VERSION.SDK_INT >= 23) {
            getWindow()!!.getDecorView()!!.setSystemUiVisibility(
                getWindow()!!.getDecorView()!!.getSystemUiVisibility() or
                    View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
            )
        }
    }

    private fun card(): LinearLayout? {
        val c: LinearLayout = LinearLayout(this)
        c.setOrientation(LinearLayout.VERTICAL)
        c.setPadding(dp(12), dp(10), dp(12), dp(10))
        c.setBackground(roundStroke(CARD, 18, BORDER, 1))
        if (Build.VERSION.SDK_INT >= 21) c.setElevation(dp(1).toFloat())
        return c
    }

    private fun detail(s: String?): TextView? {
        val t: TextView? = text(s, 12, false, MUTED)
        t!!.setLineSpacing(0f, 1.08f)
        return t
    }

    private fun text(s: String?, sp: Int, bold: Boolean, color: Int): TextView? {
        val t: TextView = TextView(this)
        t.setText(s)
        t.setTextSize(sp.toFloat())
        t.setTextColor(color)
        if (bold) t.setTypeface(Typeface.DEFAULT_BOLD)
        return t
    }

    private fun round(color: Int, radius: Int): GradientDrawable? {
        val g: GradientDrawable = GradientDrawable()
        g.setColor(color)
        g.setCornerRadius(dp(radius).toFloat())
        return g
    }

    private fun roundStroke(color: Int, radius: Int, stroke: Int, width: Int): GradientDrawable? {
        val g: GradientDrawable? = round(color, radius)
        g!!.setStroke(dp(width), stroke)
        return g
    }

    private fun margin(l: Int, t: Int, r: Int, b: Int): LinearLayout.LayoutParams? {
        val p: LinearLayout.LayoutParams = LinearLayout.LayoutParams(-1, -2)
        p.setMargins(dp(l), dp(t), dp(r), dp(b))
        return p
    }

    private fun dp(v: Int): Int {
        return Math.round(v * getResources()!!.getDisplayMetrics()!!.density)
    }

    companion object {
        private val NAVY: Int = Color.WHITE
        private val BG: Int = Color.rgb(245, 244, 240)
        private val CARD: Int = Color.WHITE
        private val TEXT: Int = Color.rgb(32, 42, 39)
        private val MUTED: Int = Color.rgb(98, 110, 104)
        private val BORDER: Int = Color.rgb(226, 229, 224)
        private val GREEN: Int = Color.rgb(40, 101, 76)
        private val ORANGE: Int = Color.rgb(197, 101, 16)
        private val LIVE_MS: Long = 10000L
        private val HISTORY_MS: Long = 24L * 60L * 60L * 1000L

        private fun onOff(v: Boolean): String? {
            return if (v) "включено" else "выключено"
        }

        private fun empty(v: String?): String? {
            return if (v == null || v!!.trim({ it <= ' ' }).isEmpty()) "—"
            else v!!.trim({ it <= ' ' })
        }

        private fun dateTime(ts: Long): String? {
            return SimpleDateFormat("dd.MM HH:mm:ss", Locale.getDefault()).format(Date(ts))
        }

        private fun age(ts: Long, now: Long): String? {
            val sec: Long = Math.max(0, (now - ts) / 1000L)
            if (sec < 60) return sec.toString() + " сек"
            val min: Long = sec / 60
            if (min < 60) return min.toString() + " мин"
            val h: Long = min / 60
            if (h < 24) return h.toString() + " ч " + (min % 60) + " мин"
            return (h / 24).toString() + " д " + (h % 24) + " ч"
        }

        private fun duration(ms: Long): String? {
            val m: Long = Math.max(0, ms / 60000L)
            val h: Long = m / 60
            val r: Long = m % 60
            return if (h > 0) h.toString() + " ч " + r.toString() + " мин"
            else m.toString() + " мин"
        }
    }
}
